### ISSUE TITLE
Performance: Aggregating by window with an aggregate is not pushed down

### ISSUE THREAD

**Issue by jsternberg** [influxdata/influxdb#14282]
A standard query where we aggregate by a window is not optimized and won't work correctly. I believe storage is capable of it, but we removed the planner rule that took advantage of this.

```
from(bucket: "system_usage")
    |> range(start: -7d)
    |> group()
    |> aggregateWindow(every: 1h, fn: count, createEmpty: true)
```

The `window` call and the subsequent `count` will not be sent down to storage so this will cause all of the data to be buffered in memory. This is definitely not intended as this specific workflow should not cause a memory allocation limit to be reached.

**Comment by jsternberg** [comment]
If the storage side does not support this, we will likely have to implement this on the client as it reads the data coming from storage. It is expected that these queries will work over long periods and will not cause a memory allocation failure.

**Comment by nathanielc** [comment]
As part of the design we do not want to push down windowing logic into the storage layer. Windowing is really just choosing start and stop bounds. As a result pushing down that logic increases the complexity of the read API but does not allow it to make any kind of optimization. As such it is not a good candidate for push down logic.

I see two ways to the planner could optimize this:

1.  Flux can determine the windowing bounds and make a request to storage for each window
2. Flux can make a single request and do the windowing client side.

Option 1 is likely to be beneficial when an aggregate can be pushed down. 

Maybe we create a planner rule to break out windows only when an aggregate can be pushed down?

**Comment by jsternberg** [comment]
I don't agree that windowing is just making multiple requests for different time ranges. I think there's actual optimization that can be done on the storage layer for windowing that isn't necessarily possible to do with that method.

At the same time, I do think there's a reasonable amount of optimization that can be done on the client side too for some more general optimizations outside of the specific storage one. Maybe we can address the specific client side optimization as option 2 first and then get to an influxdb specific optimization later. If option 1 is sufficient, then I can be wrong and things can be more simple.

And the optimization that can be done at the storage layer is to have it avoid allocating or sending back memory over the wire and that avoids any kind of marshaling or unmarshaling step between the two. But I think we both agree that a planner rule can be made for flux itself that does most of the optimization on the client side and also works for other sources outside of influxdb.

I will make another issue for windowing + aggregates and we can put this one on the icebox.

**Comment by jsternberg** [comment]
I'm going to think about it for the night and I think the above can be generalized to other narrow transformations too so I don't want to make an issue quite yet. I'm wondering if we should start seeing if we can use retaining of columns more often instead of copying.
